R/utils.R

```diff
@@ -185,3 +185,26 @@ paste_classes <- function(x) {
 is_bool <- function(x) {
   is_logical(x, n = 1) && !is.na(x)
 }
+
+friendly_element_type_of <- function(x, single = FALSE) {
+  if (is.object(x)) {
+    classes <- paste0("`", paste_classes(x), "`")
+    if (single) {
+      friendly <- sprintf("a single %s element", classes)
+    } else {
+      friendly <- sprintf("a %s element", classes)
+    }
+    return(friendly)
+  }
+
+  switch(typeof(x),
+    logical   = if (single) "a single logical"        else "a logical element",
+    integer   = if (single) "a single integer"        else "an integer element",
+    double    = if (single) "a single double"         else "a double element",
+    complex   = if (single) "a single complex number" else "a complex number",
+    character = if (single) "a single string"         else "a character vector element",
+    raw       = if (single) "a single raw value"      else "a raw vector element",
+    list      = if (single) "a single list element"   else "a list element",
```

Comment:
Is element correct here?

Reply:
What do you mean? Would you prefer something like "an R object"?

Reply:
I can't quite imagine what this would look like in an actual call. Can you give an example?

Reply:
I have simplified `friendly_type_element_of()` to always return a description for a single value:

```r
  switch(typeof(x),
    logical   = "a single logical",
    integer   = "a single integer",
    double    = "a single double",
    complex   = "a single complex number",
    character = "a single string",
    raw       = "a single raw value",
    list      = "a list of one element",
    abort("Expected a base vector type")
  )
```

Example of error messages:

```r
stop_bad_vector(1:3, character(), 1)
#> Error: Vector must be a single string, not an integer vector of length 3

stop_bad_vector(1:3, list(), 1)
#> Error: Vector must be a list of one element, not an integer vector of length 3
```